File: sdk-java/src/test/java/io/superdurable/dex/iwfcompat/NoStateFlow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.iwfcompat;

import io.superdurable.dex.Attribute;
import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.PersistenceSchema;
import io.superdurable.dex.RPC;
import io.superdurable.dex.RPCResult;

final class NoStateFlow implements Flow<Void> {
    private final Attribute<Integer> counter = Attribute.define("counter", Integer.class);

    @Override
    public PersistenceSchema getPersistenceSchema() {
        return PersistenceSchema.of(counter);
    }

    @RPC(lockAttributes = {"counter"})
    public RPCResult<Integer> increaseCounter(final Context context) {
        final int next = counter.get(context) + 1;
        counter.set(context, next);
        return RPCResult.of(next);
    }

    @RPC
    public RPCResult<Integer> getCounter(final Context context) {
        return RPCResult.of(counter.get(context));
    }

    @RPC
    public RPCResult<Long> fail(final Context context, final String input) {
        throw new IllegalArgumentException(input);
    }
}

File: sdk-typescript/test/integ/no_state_flow.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import {
  Attribute,
  StepList,
  doubleCodec,
  rpc,
  stringCodec,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
} from "../../src/index.js";

const counter = new Attribute("counter", doubleCodec);

export class NoStateFlow implements Flow {
  public readonly counter = counter;

  public getFlowType(): string {
    return "NoStateFlow";
  }

  public getSteps() {
    return StepList.empty();
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { attributes: [this.counter] };
  }

  @rpc({ outputCodec: doubleCodec, lockAttributes: [counter.lock()] })
  public increaseCounter(context: Context): RPCResult<number> {
    const next = (this.counter.get(context) ?? 0) + 1;
    this.counter.set(context, next);
    return { output: next };
  }

  @rpc({ outputCodec: doubleCodec })
  public getCounter(context: Context): RPCResult<number> {
    return { output: this.counter.get(context) };
  }

  @rpc({ inputCodec: stringCodec, outputCodec: doubleCodec })
  public fail(_context: Context, input: string): RPCResult<number> {
    throw new Error(input);
  }

  @rpc({ inputCodec: stringCodec, outputCodec: doubleCodec })
  public invoke(context: Context, _input: string): RPCResult<number> {
    if (context.flowId === "" || context.runId === "") {
      throw new Error("invalid RPC context");
    }
    return { output: 100 };
  }
}
